bin/translate-data.ts

```diff
@@ -0,0 +1,186 @@
+    this.entries = [];
+    this.children = {};
+  }
+
+  static async gatherFromDir(sourcePath: string): Promise<DataGroup> {
+    const gathered = new DataGroup();
+    gathered.sourcePath = sourcePath;
+
+    for (const entry of await fs.readdir(sourcePath, { encoding: "utf-8", withFileTypes: true })) {
+      const fullPath = path.join(sourcePath, entry.name);
+      const { base: nameWithExtension } = path.parse(fullPath);
+
+      if (entry.isDirectory()) {
+        gathered.children[nameWithExtension] = await DataGroup.gatherFromDir(fullPath);
+      } else if (entry.isFile()) {
+        if (nameWithExtension == "_type") {
+          const newTypeAnnotation = await this.loadTypeFromFile(fullPath);
+          if (gathered.typeAnnotation && gathered.typeAnnotation != newTypeAnnotation) {
+            throw new Error(
+              `Conflicting type annotation for group at ${sourcePath}: had ` +
+                `${gathered.typeAnnotation} but found ${newTypeAnnotation} ` +
+                `from ${fullPath}`,
+            );
+          }
+          gathered.typeAnnotation = newTypeAnnotation;
+          continue;
+        }
+
+        gathered.entries.push(await DataEntry.fromFile(fullPath, gathered));
+      }
+    }
+
+    return gathered;
+  }
+
+  async writeToDir(destPath: string): Promise<void> {
+    await fs.mkdir(destPath, { recursive: true });
+
+    for (const entry of this.entries) {
+      await entry.writeIntoDir(destPath);
+    }
+
+    for (const [name, group] of Object.entries(this.children)) {
+      const subdir = path.join(destPath, name);
+      await group.writeToDir(subdir);
+    }
+  }
+
+  private static async loadTypeFromFile(filePath: string): Promise<string> {
+    const typeAnnotation = (await fs.readFile(filePath, { encoding: "utf-8" })).trim();
+    assert(typeAnnotation.length > 0);
+    return typeAnnotation;
+  }
+
+  async getExpectedType(): Promise<string | null> {
+    if (this.typeAnnotation) {
+      return this.typeAnnotation;
+    }
+    if (this.sourcePath) {
+      this.typeAnnotation = await DataGroup.loadTypeFromFile(path.join(this.sourcePath, "_type"));
+    }
+    assert(this.typeAnnotation);
+    return this.typeAnnotation;
+  }
+}
+
+interface NimbusDataEntry {
+  [key: string]: unknown;
+  __nimbusMeta?: {
+    type?: string;
+  };
+}
+
+class DataEntry {
+  name: string;
+  group: DataGroup;
+  contents: unknown;
+
+  constructor(name: string, group: DataGroup, contents: unknown) {
+    this.name = name;
+    this.group = group;
+    this.contents = contents;
+  }
+
+  static async fromFile(fullPath: string, group: DataGroup): Promise<DataEntry> {
+    const { name: nameWithoutExtension, ext } = path.parse(fullPath);
+
+    let parsedContents: NimbusDataEntry;
+    let suppressTypeCheck = false;
+
+    switch (ext) {
+      case ".json": {
+        try {
+          const fileContents = await fs.readFile(fullPath, { encoding: "utf-8" });
+          parsedContents = JSON.parse(fileContents);
```

Comment:
Q: Can we just import the JSON directly (similar to below for '.ts' files or using `require("./package.json")` in vanilla node?

```ts
parsedContents = (await import(`../${fullPath}`)).default;
```

Reply:
Typescript doesn't like importing JSON files directly. You can do it, but I don't see the benefit of it.

Reply:
I think this implementation makes sense as-is